test/unit/org/apache/cassandra/db/virtual/GossipStateTransitionsTableTest.java

```diff
@@ -0,0 +1,96 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.virtual;
+
+import com.google.common.collect.ImmutableList;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.CQLTester;
+import org.apache.cassandra.cql3.UntypedResultSet;
+import org.apache.cassandra.gms.Gossiper;
+import org.apache.cassandra.utils.FBUtilities;
+
+import static org.apache.cassandra.utils.Clock.Global.currentTimeMillis;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for the {@link GossipStateTransitionsTable} virtual table.
+ */
+public class GossipStateTransitionsTableTest extends CQLTester
+{
+    private static final String KS_NAME = "vts";
+
+    @SuppressWarnings("FieldCanBeLocal")
+    private VirtualTable table;
+
+    @BeforeClass
+    public static void setup() throws Exception
+    {
+        // limit the transitions to record to 2 per host
+        DatabaseDescriptor.setMaxGossipStateTransitionsSize(2);
+        setUpClass();
+        startJMXServer();
+    }
+
+    @Before
+    public void config()
+    {
+        table = new GossipStateTransitionsTable(KS_NAME);
+        VirtualKeyspaceRegistry.instance.register(new VirtualKeyspace(KS_NAME, ImmutableList.of(table)));
+    }
+
+    @Test
+    public void testSelectAllWhenNoGossipStateTransitionsHaveBeenRecorded() throws Throwable
+    {
+        assertEmpty(execute("SELECT * FROM vts.gossip_state_transitions"));
+    }
+
+    @Test
+    public void testSelectAllWithStateTransitions() throws Throwable
+    {
+        long testStartMillis = currentTimeMillis();
+        try
+        {
+            requireNetwork(); // triggers gossip state transitions
+
+            UntypedResultSet resultSet = execute("SELECT * FROM vts.gossip_state_transitions");
+
+            assertThat(resultSet.size()).isGreaterThan(0)
+                                        .isLessThanOrEqualTo(2);
+
+            for (UntypedResultSet.Row row : resultSet)
+            {
+                assertThat(row.getColumns().size()).isEqualTo(24);
+                assertThat(row.getTimestamp("timestamp").getTime()).isGreaterThan(testStartMillis)
+                                                                   .isLessThan(currentTimeMillis());
```

Comment:
this can be flaky as millis is free to go backwards (yay clocks!), and also free to not change.

Reply:
I believe `nanoTime()` can go backwards; millis is guaranteed to at least not go backwards. I actually forget which one is guaranteed. But I agree that this is risky, so I will think about something for this

Reply:
`nano` can't go backwards its based off cpu cycles, `millis` can and does go backwards.  A simple way to trigger that is ntp